### PR TITLE
feat: create cloud backup

### PR DESCRIPTION
**Pull request by Glitchfix**
CSI create snapshot will now support triggering snapshot that will initiate a cloud backup.

This only support with default cloud backup credentials to be used to create the backups with the same name as the snapshot.

**Special notes for your reviewer**:
This PR additionally proposes to try a different way of unit test.

### CODE REVIEW

File: csi/controller.go
Comment (ggriffiths) [inline csi/controller.go:1014]: We'll need to use `getRemoteConn` to spread the load of cloud backups across all nodes. As a result, it's best not to store this client locally in OsdCsiServer.

See this line for example:
https://github.com/libopenstorage/openstorage/blob/master/csi/controller.go#L548-L557

File: csi/controller.go
Comment (ggriffiths) [inline csi/controller.go:1067]: remove Printf

File: csi/controller.go
Comment (ggriffiths) [inline csi/controller.go:902]: Can you externalize this to be a constant?

Also, let's name this with our standard convention: `csi.openstoarge.org/snapshot-type`.

See this area for an example of the prefix:
https://github.com/libopenstorage/openstorage/blob/master/csi/controller.go#L42-L46

File: csi/controller.go
Comment (ggriffiths) [inline csi/controller.go:907]: Values should be `local` or `cloud`

File: csi/controller.go
Comment (ggriffiths) [inline csi/controller.go:1056]:
```suggestion
		return nil, status.Errorf(codes.Aborted, "Failed to create cloud snapshot: %v", err)
```

File: csi/controller.go
Comment (ggriffiths) [inline csi/controller.go:1054]: This was a special case for local snapshots. We can remove this from cloud snapshots.

File: csi/controller.go
Comment (ggriffiths) [inline csi/controller.go:1041]: If we a return a non-final gRPC error code after this line, will a subsequent s.cloudBackupClient.Create cause issues here? Did you test how that workflow would work?

File: csi/controller.go
Comment (ggriffiths) [inline csi/controller.go:1076]: Will this cause an infinite cycle of snapshot delete/create? We want to return a non-final error let the snapshotter retry in this case.

File: csi/controller_test.go
Comment (ggriffiths) [inline csi/controller_test.go:3391]: in our call, you mentioned this needs to be a part of the Csi struct in order for this to work.

However, we'll need to find a way to generate this client per test. The reason is that we need a new client for each remote conn we get for round-robin connections.

Comment (ggriffiths) [inline csi/controller_test.go:3391]: in our call, you mentioned this needs to be a part of the Csi struct in order for this to work.

However, we'll need to find a way to generate this client per test. The reason is that we need a new client for each remote conn we get for round-robin connections.

Comment (ggriffiths) [inline csi/controller_test.go:3391]: in our call, you mentioned this needs to be a part of the Csi struct in order for this to work.

However, we'll need to find a way to generate this client per test. The reason is that we need a new client for each remote conn we get for round-robin connections.

File: csi/controller.go
Comment (Glitchfix) [reply csi/controller.go:1076]: this could happen, or it could generate redundant copies

**Comment by ggriffiths** [comment]
@Glitchfix can you fix the CSI unit tests? They are failing

File: csi/controller.go
Comment (ggriffiths) [reply csi/controller.go]: Can you test this implementation in a live PX cluster? I think this will continuously create new snapshots on every retry if it does not exit immediately.

**Comment by Glitchfix** [comment]
> @Glitchfix can you fix the CSI unit tests? They are failing

There was a typo that occurred when I change the label which is why it was failing

File: csi/controller.go
Comment (adityadani) [inline csi/controller.go:1048]: The CloudBackupCreate request also takes in "CredentialID". We should take that as an input from the user.  Similar to `snapshot-type`, lets take an input `credential-id`

File: csi/controller.go
Comment (adityadani) [inline csi/controller.go:1093]: You are using a deprecated size field from the response
```
message SdkCloudBackupSizeResponse{
  // Same value as total_download_bytes, will be deprecated
  uint64 size = 1;
  // Total download size of all the backups (full & incremental) required for a restore
  uint64 total_download_bytes = 2;
  // Size of the compressed individual backup in object-store or NFS
  uint64 compressed_object_bytes = 3;
  // Used size of the volume to be restored (in bytes)
  uint64 capacity_required_for_restore = 4;
}
```
According to CSI spec, what is the expected value from the size field, since it needs to map to one of these values.

File: csi/controller.go
Comment (adityadani) [inline csi/controller.go:1097]: Don't we need to change this flag to `true` at some point in time? 
I think previously our capability was only for LocalSnapshots which were instantaneous. We might have to change that since cloud backups take some time. 

Check this comment on ListSnapshots
```
// ListSnapshots lists all snapshots in a cluster.
// This is mainly implemented for Nomad, as Kubernetes will not call
// list snapshots for drivers which have synchronous snapshot creation.
// This is because ReadyToUse is set to true immediately upon on our snapshot creation.
```
cc : @ggriffiths

Comment (Glitchfix) [reply csi/controller.go:1100]: I see an implementation in list snapshot where the snapshot status is fetched first before returning it to the caller of the API.
for cloud backups, we could first set it to a pending state and fetch the status of the snaps whenever the list APIs are invoked.
I will raise that change in the list snapshot PR

File: csi/controller.go
Comment (adityadani) [reply csi/controller.go:1096]: Are you planning to fix this?

File: csi/controller.go
Comment (adityadani) [reply csi/controller.go:1100]: But right now if the backup status is Active you will set this to `true`. Before we merge this change we should fix that. Only set isBackupReady to `true` iff status is SdkCloudBackupStatusType_SdkCloudBackupStatusTypeDone

**Review by adityadani — COMMENTED** [review]
We need to handle the two outstanding comments before merging this change.

**Review by ggriffiths — APPROVED** [review]
Lgtm!

**Comment by ggriffiths** [comment]
@Glitchfix I ran travis again, and it looks like a snapshot test is failing: 
https://app.travis-ci.com/github/libopenstorage/openstorage/jobs/605718589#L1559

Can we run this test a handful of times and see if it's flakey?

**Comment by Glitchfix** [comment]
> @Glitchfix I ran travis again, and it looks like a snapshot test is failing: https://app.travis-ci.com/github/libopenstorage/openstorage/jobs/605718589#L1559
> 
> Can we run this test a handful of times and see if it's flakey?

there was one deprecated test case that was being executed I think it got added back somehow while rebasing
I have removed it now

**Comment by Glitchfix** [comment]
@ggriffiths we can now merge it
I don't have the write access to merge it, need help

**Pull request merged**
